File: windows/tests/Elevate.App.Tests/AppModelSharedAppTests.cs
using System.Web;
using Elevate.App.Services;
using Elevate.App.Tests.Support;
using Elevate.App.ViewModels;
using Elevate.Core.Auth;
using Elevate.Core.Models;
using FluentAssertions;

namespace Elevate.App.Tests;

/// <summary>
/// Port of the macOS <c>AppModelSharedAppTests</c>: <see cref="AppSettings.UsesSharedClientId"/>,
/// <see cref="AppModel.UsesSharedApp"/> and <see cref="AppModel.SharedAppAdminConsentUrl"/>, and the
/// diagnostics line that names the shared app without ever printing the id itself.
/// </summary>
public class AppModelSharedAppTests
{
    private const string OtherClientId = "11111111-2222-3333-4444-555555555555";

    private static string Query(Uri url, string name) => HttpUtility.ParseQueryString(url.Query)[name]!;

    [Fact]
    public void UsesSharedClientIdMatchesAnyCaseWithWhitespace()
    {
        using var test = new TestModel(clientId: $"  {AppSettings.SharedClientId.ToUpperInvariant()}  ");
        test.Settings.UsesSharedClientId.Should().BeTrue();
    }

    [Fact]
    public void UsesSharedClientIdFalseForAnotherGuidOrEmpty()
    {
        using var other = new TestModel(clientId: OtherClientId);
        other.Settings.UsesSharedClientId.Should().BeFalse();
        using var empty = new TestModel(clientId: "");
        empty.Settings.UsesSharedClientId.Should().BeFalse();
    }

    [Fact]
    public async Task SharedAppAdminConsentUrlIsNullWhenUnconfigured()
    {
        using var test = await TestModel.BootstrappedAsync(ownApp: new FakeOwnAppProvider(), clientId: "");
        test.Model.IsConfigured.Should().BeFalse();
        test.Model.SharedAppAdminConsentUrl().Should().BeNull();
    }

    [Fact]
    public async Task SharedAppAdminConsentUrlWhenConfiguredWithSharedId()
    {
        using var test = await TestModel.BootstrappedAsync(ownApp: new FakeOwnAppProvider(), clientId: AppSettings.SharedClientId);
        test.Model.IsConfigured.Should().BeTrue();
        test.Model.UsesSharedApp.Should().BeTrue();

        var url = test.Model.SharedAppAdminConsentUrl();

        url.Should().NotBeNull();
        url!.Host.Should().Be("login.microsoftonline.com");
        url.AbsolutePath.Should().Be("/organizations/v2.0/adminconsent");
        Query(url, "client_id").Should().Be(AppSettings.SharedClientId);
        Query(url, "redirect_uri").Should().Be(AppSettings.SharedConsentRedirectUri);
        Query(url, "scope").Should().Contain("RoleAssignmentSchedule.ReadWrite.Directory");
    }

    [Fact]
    public async Task SharedAppAdminConsentUrlIsNullForAnotherGuid()
    {
        using var test = await TestModel.BootstrappedAsync(ownApp: new FakeOwnAppProvider(), clientId: OtherClientId);
        test.Model.IsConfigured.Should().BeTrue();
        test.Model.UsesSharedApp.Should().BeFalse();
        test.Model.SharedAppAdminConsentUrl().Should().BeNull();
    }

    [Fact]
    public async Task PerTenantAdminConsentUrlUsesTenantIdAndConsentPageRedirectForSharedApp()
    {
        using var test = await TestModel.BootstrappedAsync(ownApp: new FakeOwnAppProvider(), clientId: AppSettings.SharedClientId);
        test.Model.State.Identities.Add(Sample.Identity(method: SignInMethod.OwnApp));

        var url = test.Model.AdminConsentUrl(Sample.IdentityId, Sample.TenantId);

        url.Should().NotBeNull();
        url!.AbsolutePath.Should().Be($"/{Sample.TenantId}/v2.0/adminconsent");
        Query(url, "redirect_uri").Should().Be(AppSettings.SharedConsentRedirectUri);
    }

    [Fact]
    public async Task PerTenantAdminConsentUrlKeepsNativeClientRedirectForOwnRegistration()
    {
        using var test = await TestModel.BootstrappedAsync(ownApp: new FakeOwnAppProvider(), clientId: OtherClientId);
        test.Model.State.Identities.Add(Sample.Identity(method: SignInMethod.OwnApp));

        var url = test.Model.AdminConsentUrl(Sample.IdentityId, Sample.TenantId);

        url.Should().NotBeNull();
        Query(url!, "redirect_uri").Should().Be(SharedApp.NativeClientRedirectUri);
    }

    [Fact]
    public async Task ApplyingTheSharedIdMakesItTheOneInEffect()
    {
        using var test = await TestModel.BootstrappedAsync(
            ownApp: new FakeOwnAppProvider(), ownAppFactory: _ => new FakeOwnAppProvider(), clientId: OtherClientId);

        test.Model.ApplyClientId(AppSettings.SharedClientId);

        test.Model.UsesSharedApp.Should().BeTrue();
        test.Model.SharedAppAdminConsentUrl().Should().NotBeNull();
    }

    [Fact]
    public async Task DiagnosticsNamesSharedAppWithoutTheIdItself()
    {
        using var test = await TestModel.BootstrappedAsync(ownApp: new FakeOwnAppProvider(), clientId: AppSettings.SharedClientId);
        var text = test.Model.DiagnosticsText();
        text.Should().Contain("Client id: shared Elevate app").And.NotContain(AppSettings.SharedClientId);
    }

    [Fact]
    public async Task DiagnosticsShowsOwnRegistrationForOtherClientId()
    {
        using var test = await TestModel.BootstrappedAsync(ownApp: new FakeOwnAppProvider(), clientId: OtherClientId);
        var text = test.Model.DiagnosticsText();
        text.Should().Contain("Client id: own registration")
            .And.NotContain("shared Elevate app").And.NotContain("not set").And.NotContain(OtherClientId);
    }

    [Fact]
    public async Task DiagnosticsShowsNotSetWhenUnconfigured()
    {
        using var test = await TestModel.BootstrappedAsync(ownApp: new FakeOwnAppProvider(), clientId: "");
        var text = test.Model.DiagnosticsText();
        text.Should().Contain("Client id: not set").And.NotContain("own registration").And.NotContain("shared Elevate app");
    }
}

File: macos/Tests/ElevateAppTests/AppModelSharedAppTests.swift
import Foundation
import Testing
import ElevateCore
@testable import Elevate

/// The project-provided shared Elevate app registration: `AppSettings.usesSharedClientId`,
/// `AppModel.usesSharedApp`/`sharedAppAdminConsentURL()`, and the diagnostics line that names it
/// without ever printing the id itself.
@MainActor
struct AppModelSharedAppTests {
    @Test func usesSharedClientIdMatchesAnyCaseWithWhitespace() {
        let settings = makeSettings()
        settings.clientId = "  \(AppSettings.sharedClientId.uppercased())  "
        #expect(settings.usesSharedClientId)
    }

    @Test func usesSharedClientIdFalseForAnotherGUID() {
        let settings = makeSettings()
        settings.clientId = "11111111-2222-3333-4444-555555555555"
        #expect(!settings.usesSharedClientId)
    }

    @Test func usesSharedClientIdFalseForEmpty() {
        let settings = makeSettings()
        settings.clientId = ""
        #expect(!settings.usesSharedClientId)
    }

    @Test func sharedAppAdminConsentURLIsNilWhenUnconfigured() async {
        let settings = makeSettings()
        // Explicitly blank: `AppSettings` migrates a client id from the legacy defaults suite when
        // it finds one, so a fresh test suite is not necessarily unconfigured.
        settings.clientId = ""
        let model = await makeModel(settings: settings, ownAppViaLoopback: true)
        #expect(!model.isConfigured)
        #expect(model.sharedAppAdminConsentURL() == nil)
        cleanup(model)
    }

    @Test func sharedAppAdminConsentURLWhenConfiguredWithSharedId() async {
        let settings = makeSettings()
        settings.clientId = AppSettings.sharedClientId
        let model = await makeModel(settings: settings, ownAppViaLoopback: true)
        #expect(model.isConfigured)
        #expect(model.usesSharedApp)
        let url = model.sharedAppAdminConsentURL()
        #expect(url != nil)
        let components = URLComponents(url: url!, resolvingAgainstBaseURL: false)
        #expect(components?.host == "login.microsoftonline.com")
        #expect(components?.path == "/organizations/v2.0/adminconsent")
        let items = components?.queryItems ?? []
        #expect(items.first { $0.name == "client_id" }?.value == AppSettings.sharedClientId)
        #expect(items.first { $0.name == "redirect_uri" }?.value == AppSettings.sharedConsentRedirectURI)
        let scope = items.first { $0.name == "scope" }?.value ?? ""
        #expect(scope.contains("RoleAssignmentSchedule.ReadWrite.Directory"))
        cleanup(model)
    }

    @Test func sharedAppAdminConsentURLIsNilWhenConfiguredWithAnotherGUID() async {
        let settings = makeSettings()
        settings.clientId = "11111111-2222-3333-4444-555555555555"
        let model = await makeModel(settings: settings, ownAppViaLoopback: true)
        #expect(model.isConfigured)
        #expect(!model.usesSharedApp)
        #expect(model.sharedAppAdminConsentURL() == nil)
        cleanup(model)
    }

    @Test func adminConsentURLStillUsesTenantIdInPath() async {
        let settings = makeSettings()
        settings.clientId = AppSettings.sharedClientId
        let model = await makeModel(settings: settings, ownAppViaLoopback: true)
        model.state.identities = [Sample.identity(method: .ownApp)]
        let url = model.adminConsentURL(identityId: Sample.identityId, tenantId: Sample.tenantId)
        #expect(url != nil)
        #expect(URLComponents(url: url!, resolvingAgainstBaseURL: false)?.path == "/\(Sample.tenantId)/v2.0/adminconsent")
        cleanup(model)
    }

    @Test func sharedAppAdminConsentURLUsesConsentPageRedirect() async {
        let settings = makeSettings()
        settings.clientId = AppSettings.sharedClientId
        let model = await makeModel(settings: settings, ownAppViaLoopback: true)
        let url = model.sharedAppAdminConsentURL()
        #expect(url != nil)
        let items = URLComponents(url: url!, resolvingAgainstBaseURL: false)?.queryItems ?? []
        #expect(items.first { $0.name == "redirect_uri" }?.value == AppSettings.sharedConsentRedirectURI)
        cleanup(model)
    }

    @Test func adminConsentURLUsesConsentPageRedirectForSharedApp() async {
        let settings = makeSettings()
        settings.clientId = AppSettings.sharedClientId
        let model = await makeModel(settings: settings, ownAppViaLoopback: true)
        model.state.identities = [Sample.identity(method: .ownApp)]
        let url = model.adminConsentURL(identityId: Sample.identityId, tenantId: Sample.tenantId)
        #expect(url != nil)
        let items = URLComponents(url: url!, resolvingAgainstBaseURL: false)?.queryItems ?? []
        #expect(items.first { $0.name == "redirect_uri" }?.value == AppSettings.sharedConsentRedirectURI)
        cleanup(model)
    }

    @Test func adminConsentURLKeepsNativeClientRedirectForOwnRegistration() async {
        let settings = makeSettings()
        settings.clientId = "11111111-2222-3333-4444-555555555555"
        let model = await makeModel(settings: settings, ownAppViaLoopback: true)
        model.state.identities = [Sample.identity(method: .ownApp)]
        #expect(!model.usesSharedApp)
        let url = model.adminConsentURL(identityId: Sample.identityId, tenantId: Sample.tenantId)
        #expect(url != nil)
        let items = URLComponents(url: url!, resolvingAgainstBaseURL: false)?.queryItems ?? []
        #expect(items.first { $0.name == "redirect_uri" }?.value == "https://login.microsoftonline.com/common/oauth2/nativeclient")
        cleanup(model)
    }

    @Test func diagnosticsTextNamesSharedAppWithoutTheIdItself() async {
        let settings = makeSettings()
        settings.clientId = AppSettings.sharedClientId
        let model = await makeModel(settings: settings, ownAppViaLoopback: true)
        let text = model.diagnosticsText()
        #expect(text.contains("Client id: shared Elevate app"))
        #expect(!text.contains(AppSettings.sharedClientId))
        cleanup(model)
    }

    @Test func diagnosticsTextShowsOwnRegistrationForOtherClientId() async {
        let settings = makeSettings()
        settings.clientId = "11111111-2222-3333-4444-555555555555"
        let model = await makeModel(settings: settings, ownAppViaLoopback: true)
        let text = model.diagnosticsText()
        #expect(!text.contains("Client id: shared Elevate app"))
        #expect(text.contains("Client id: own registration"))
        #expect(!text.contains("Client id: not set"))
        cleanup(model)
    }

    @Test func diagnosticsTextShowsNotSetWhenUnconfigured() async {
        let settings = makeSettings()
        settings.clientId = ""
        let model = await makeModel(settings: settings, ownAppViaLoopback: true)
        #expect(!model.isConfigured)
        let text = model.diagnosticsText()
        #expect(text.contains("Client id: not set"))
        #expect(!text.contains("Client id: shared Elevate app"))
        #expect(!text.contains("Client id: own registration"))
        cleanup(model)
    }
}
